Prox/Prox/Utilities/GoogleDirectionsMatrixTravelTimesProvider.swift

```diff
@@ -43,8 +46,71 @@ class GoogleDirectionsMatrixTravelTimesProvider: TravelTimesProvider {
+        var fetchSize = min(locations.endIndex, 25) - 1
+        var expectedResponseCount = 0
+        var allTravelTimes = [TravelTimes]()
+        let deferred = Deferred<DatabaseResult<[TravelTimes]>>()
+        while index < locations.endIndex {
+            let subarray = Array(locations[index...fetchSize])
+
+            if let url = distanceMatrixURL(fromLocation: fromLocation, toLocations: subarray, byTransportType: mapTransportTypeToMode(transportType: transitType)) {
+                let request = URLRequest(url: url, cachePolicy: URLRequest.CachePolicy.returnCacheDataElseLoad, timeoutInterval: 2 * AppConstants.ONE_MINUTE)
+                let dataTask = sessionManager.dataTask(with: request) { response, object, error in
+                    expectedResponseCount -= 1
+                    if error != nil {
+                        NSLog("Error fetching Google Distance Matrix request \(error)")
+                        return
+                    }
+
+                    guard let responseObject = object as? [String: Any],
+                        let row = (responseObject["rows"] as? [[String: Any]])?.first,
+                        let elements = row["elements"] as? [[String: Any]]
+                        else {
+                            NSLog("Error parsing Google Distance Matrix Response \(object)")
+                            return
+                    }
+                    for (index, result) in elements.enumerated() {
+                        guard index < subarray.count,
+                        let duration = result["duration"] as? [String: Any],
+                        let value = duration["value"] as? Double else {
+                            continue
+                        }
+
+                        let latLon = subarray[index]
+                        let key = (toLocations as NSDictionary).allKeys(for: latLon).first as? PlaceKey
+
+                        let travelTime: TravelTimes
+                        switch transitType {
+                        case MKDirectionsTransportType.automobile:
+                            travelTime = TravelTimes(origin: fromLocation, destination: latLon, destinationPlaceKey: key, walkingTime: nil, drivingTime: value, publicTransportTime: nil)
+                        case MKDirectionsTransportType.transit:
+                            travelTime = TravelTimes(origin: fromLocation, destination: latLon, destinationPlaceKey: key, walkingTime: nil, drivingTime: nil, publicTransportTime: value)
+                        default:
+                            travelTime = TravelTimes(origin: fromLocation, destination: latLon, destinationPlaceKey: key, walkingTime: value, drivingTime: nil, publicTransportTime: nil)
+                        }
+                        allTravelTimes.append(travelTime)
+                    }
+                    if expectedResponseCount == 0 {
+                        deferred.fill(with: DatabaseResult.succeed(value: allTravelTimes))
```

Comment:
I'm not sure which thread data task runs on (the session manager's but assuming it's concurrent, I'm concerned this can be called early. We increment `expectedResponseCount`, decrement it (in the data task response), and check `== 0` before the while loop iterates again.

It looks like we can calculate expectedResponseCount in advance – perhaps we should do that? (unless you know the response handler is called on the same thread